File: csharp/332.重新安排行程.cs
/*
 * @lc app=leetcode.cn id=332 lang=csharp
 *
 * [332] 重新安排行程
 */

// @lc code=start
public class Solution
{
    public IList<string> FindItinerary(IList<IList<string>> tickets)
    {
        // build tree
        var adj = new Dictionary<string, List<List<string>>>();
        unvisitedCnt = tickets.Count;
        visited = new Dictionary<List<string>, bool>();
        foreach (var t in tickets)
        {
            string from = t[0];
            if (!adj.ContainsKey(from))
                adj.Add(from, new List<List<string>>());
            adj[from].Add(t);
            visited.Add(t, false);
        }

        // Find paths
        TraversePath(adj, "JFK", new List<string>() { "JFK" });
        return MinPath(paths);
    }

    List<List<string>> paths = new List<List<string>>();
    Dictionary<List<string>, bool> visited;
    int unvisitedCnt = 0;
    void TraversePath(Dictionary<string, List<List<string>>> adj, string from, List<string> path)
    {
        // base case
        if (unvisitedCnt == 0)
        {
            path.Add(new List<string>(path));
            return;
        }

        // take one path
        foreach (var ticket in adj[from])
        {
            if (visited[ticket]) continue;
            visited[ticket] = true;
            unvisitedCnt--;
            path.Add(ticket[1]);
            TraversePath(adj, ticket[1], path);
            visited[ticket] = false;
            unvisitedCnt++;
        }
    }

    List<string> MinPath(List<List<string>> paths)
    {

    }
}
// @lc code=end

File: java/332.重新安排行程.java
import java.awt.List;
import java.util.Collection;
import java.util.Collections;
import java.util.Map;
import java.util.PriorityQueue;

/*
 * @lc app=leetcode.cn id=332 lang=java
 *
 * [332] 重新安排行程
 */

// @lc code=start
class Solution {
    List<String> itinerary = new LinkedList<String>();
    Map<String, PriorityQueue<String>> map = new HashMap<String, PriorityQueue<String>>();

    public List<String> findItinerary(List<List<String>> tickets) {
        // build tree
        for (List<String> ticket : tickets) {
            String src = ticket.get(0);
            String dst = ticket.get(1);
            if (!map.containsKey(src))
                map.put(src, new PriorityQueue<String>());
            map.get(src).offer(dst);
        }

        // do dfs
        dfs("JFK");
        Collections.reverse(itinerary);
        return itinerary;
    }

    void dfs(String curr) {
        while (map.containsKey(curr) && map.get(curr).size() > 0) {
            dfs(map.get(curr).poll());
        }
        itinerary.add(curr);
    }
}
// @lc code=end
